feature/footer-component

andrewlubrino opened the pull request:
**Pull request recommendations:**
- [x] Name your pull request _your-development-type/short-description_. Ex: _feature/read-tiff-files_
- [x] Link to any relevant issue in the PR description. Ex: _Resolves [gh-12], adds tiff file format support_

Resolves #36 

- [x] Provide context of changes.

Adds reusable footer component w/ parametrizable links for each instance.

- [ ] Provide relevant tests for your feature or bug fix.
- [x] Provide or update documentation for any feature added by your pull request.

Thanks for contributing!

JacksonMaxfield reviewed (changes requested):
Hey @andrewlubrino this is really great so far! I think most of my comments are about "PR Management" and some nitpicks / asks:

1. I think you have an odd case because you have two open PRs (which is **great**) but in the future whenever possible can you work off of your `main` branch so that you can drop a link to your storybook docs, specifically a link to the component you're adding, updating, etc. (No worries though, we should have added that in contributing docs somewhere / figure out a better process for that. For now I just pulled your branch and generated storybook docs locally.)
2. I updated the PR title + description just to alert ya.
3. @hawkticehurst @tohuynh should this have Storybook controls? And I will leave it to you two to determine if it needs tests too (or _what_ to test)? Sorry for my JS lack of knowledge.

Thanks again, @andrewlubrino! Love to see this work :100:

hawkticehurst commented on src/components/Layout/Footer/FooterLayout.tsx:
```diff
@@ -0,0 +1,110 @@
+  return (
+    <footer className="mzp-c-footer">
+      <div className="mzp-l-content">
+        <section className="mzp-c-footer-section">
+          <h1 className="mzp-c-footer-heading" style={{ fontSize: "1.65rem" }}>
+            Council Data Project
+          </h1>
+          <p>{about}</p>
+          <form action={aboutLinkDestination}>
+            <input
+              className="mzp-c-button mzp-t-product"
+              type="submit"
+              value="About the project"
+              style={{
+                borderRadius: ".25rem",
+                border: "none",
+                height: "2.25rem",
+                backgroundColor: "#307D7E",
+                color: "white",
+                fontWeight: "bold",
+              }}
+            />
+          </form>
+        </section>
```
Yeah, I agree. I like the idea but since people will very likely see all of this information elsewhere I don't it necessarily needs to be repeated in the footer.

hawkticehurst commented:
> 3\. @hawkticehurst @tohuynh should this have Storybook controls?

Theoretically, it _should_ use controls but since I haven't converted the other components yet I think you could probably just leave it for now and I'll do it all soon.

hawkticehurst reviewed (commented):
Looking pretty good! A lot of my feedback is just centered around making sure we are following Mozilla Protocol documentation/examples and making sure there's consistent styling across all the sections of the footer.

JacksonMaxfield commented on src/components/Layout/Footer/FooterLayout.tsx:
```diff
@@ -0,0 +1,110 @@
+  );
+}
+
+const FooterLayout: FC<FooterLayoutProps> = ({
```
Yea I think having the directory of components like this be called layout is fine but the component itself is fine to just be `Footer`

tohuynh replied:
Agree

JacksonMaxfield commented on src/components/Layout/Footer/FooterLayout.tsx:
```diff
@@ -0,0 +1,110 @@
+          })}
+        </section>
+        <section className="mzp-c-footer-section">
+          <h1 className="mzp-c-footer-heading">DISCLAIMER</h1>
```
^^^ Definitely. And in that way we may want to have the current disclaimer text (the somewhat legal disclaimer of "things may not be 100% accurate") +

```
Portions of this content are ©2017–2021 by individual Council Data Project contributors. Content available under MIT License.

Styled using Mozilla Protocol. Artwork provided by unDraw.
```

See the current footer of https://councildataproject.github.io

tohuynh commented on src/components/Layout/Footer/FooterLayout.tsx:
```diff
@@ -0,0 +1,110 @@
+            Council Data Project
+          </h1>
+          <p>{about}</p>
+          <form action={aboutLinkDestination}>
+            <input
+              className="mzp-c-button mzp-t-product"
+              type="submit"
+              value="About the project"
+              style={{
+                borderRadius: ".25rem",
+                border: "none",
+                height: "2.25rem",
+                backgroundColor: "#307D7E",
+                color: "white",
+                fontWeight: "bold",
+              }}
+            />
+          </form>
```
Agree about not using `form` for this, I've only seen form action for internal API endpoints.

tohuynh commented on src/components/Layout/Footer/FooterLayout.tsx:
```diff
@@ -0,0 +1,110 @@
+  aboutLinkDestination: string;
+}
+
+function renderHeader(header: Header) {
```
Could you add comments to each field of FooterLayoutProps so that we get nice documentation of the props in Storybook docs panel?

tohuynh commented on src/components/Layout/Footer/FooterLayout.tsx:
```diff
@@ -0,0 +1,110 @@
+  url: string;
+}
+
+interface Header {
```
Good idea. I like changing `Header` to `FooterLinksSection` (contains headerName and a bunch of links) and `Link` to `FooterLink`

tohuynh reviewed (commented):
Thanks for doing this. I have few comments.

In general for css-in-jss styling, if styling is specific to component you're working on, could you use emotion/styled to create the styled component? Here's an example (in PersonCard.tsx): 
```
const PersonStatus = styled.div({
  float: "right",
  marginRight: 8,
  marginBottom: 8,
  borderRadius: "10%",
  padding: "0 4px",
});
```
Then you can use the styled component in the jsx.

tohuynh commented:
>@hawkticehurst @tohuynh should this have Storybook controls? And I will leave it to you two to determine if it needs tests too (or what to test)? Sorry for my JS lack of knowledge.

Yes on storybook controls, but as Hawk said it isn't needed for this PR.

I think there's no need to test this footer component.

JacksonMaxfield commented:
After merging the other PR @andrewlubrino can you update this branch by merging the changes in and shoving this branch's storybook to gh-pages.

andrewlubrino commented:
> After merging the other PR @andrewlubrino can you update this branch by merging the changes in and shoving this branch's storybook to gh-pages.

Whoops, totally meant to make it to the meeting, but set an alarm for 11 instead of 10. Time zones always get me. I'll be sure to make the next one.

Yep, I'll pull in the changes that were just merged into main into my footer and then send the footer as-is to gh-pages.

andrewlubrino commented:
> After merging the other PR @andrewlubrino can you update this branch by merging the changes in and shoving this branch's storybook to gh-pages.

@JacksonMaxfield Just updated gh-pages. [Link](https://andrewlubrino.github.io/cdp-frontend/?path=/story/library-layout-footer--footer) to my gh-pages.

JacksonMaxfield commented:
> > After merging the other PR @andrewlubrino can you update this branch by merging the changes in and shoving this branch's storybook to gh-pages.
> 
> @JacksonMaxfield Just updated gh-pages. [Link](https://andrewlubrino.github.io/cdp-frontend/?path=/story/library-layout-footer--footer) to my gh-pages. I haven't yet made any of the updates yet. I'll work through the changes tomorrow.

Thanks! Looks like you got some stuff to resolve still so just push updates as you go. Will re-review after ya do.

and thanks again for getting started on this + the other issue!

andrewlubrino commented:
Hi everyone. I just went through the comments and made some changes. I'm pretty sure I addressed everything, but if there's anything I missed, please let me know. 

The changes were pretty extensive. I removed the button from the footer for now. The footer component looks very similar to the example on the Mozilla Protocol pages, but retains the dynamic rendering functionality that the initial pull request had.

The updated component is up on my gh-pages. [Link to my gh-pages](https://andrewlubrino.github.io/cdp-frontend/?path=/story/library-layout-footer--footer).

hawkticehurst commented:
Thanks @andrewlubrino!!

I'll take a closer look at all of these changes tonight (or tomorrow at the latest). But at least from a high level glance at the gh-pages link, your updates look great!

hawkticehurst commented:
> I'm not sure if GitHub notifies everyone when a commit is made

Also yes, GitHub notified me of your commit. 🙂

But @-ing all of us is great/perfectly acceptable too (if anything I pay more attention when I see my name 😅)

JacksonMaxfield commented on src/components/Layout/Footer/Footer.tsx:
```diff
@@ -0,0 +1,85 @@
+              <h5 className="mzp-c-footer-heading">About</h5>
+              <ul>
+                <li>
+                  <a href="https://councildataproject.github.io/#about">
+                    CouncilDataProject Information
```
Nit: Change this to just "CouncilDataProject" and remove the pointer to `#about`.

andrewlubrino replied:
Should have caught that. I'll make sure to look at the bottom of storybook more closely in the future. 

The accessibility issue I'm getting reads `The landmark must have a unique aria-label, aria-labelledby, or title to make landmarks distinguishable`. I fixed this by adding `aria-label="footer"` as an attribute on the `nav` element (storybook provided a [link](https://dequeuniversity.com/rules/axe/4.1/landmark-unique?application=axeAPI) with more details and potential solutions just above the accessibility issue; I took this solution from that link). 

I tried adding the attributes `id="footer" role="contentinfo"`, but storybook was throwing me many errors. I think the latest commit should get rid of the issue, but let me know if you're still getting an error on your end when you look at it. Newest commit is also posted to [my gh-pages](https://andrewlubrino.github.io/cdp-frontend/)

JacksonMaxfield replied:
Beautiful. Thanks for resolving. I will will wait for To or Hawk to review but all good from my side. Whenever someone else approved will happily merge. Thanks for all your work on this :slightly_smiling_face:

JacksonMaxfield reviewed (approved):
This looks great to me! Thanks for making the changes. One minor nitpick from my side.

Also I saw on storybook that there is an accessibility violation. From some quick searching looks like it can be resolved by add an id + role to the main element? probably just `id="footer" role="contentinfo"`?

See here for ARIA info: https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA11.html

tohuynh commented on src/components/Layout/Footer/Footer.stories.mdx:
```diff
@@ -0,0 +1,30 @@
+
+<Meta title="Library/Layout/Footer" component={Footer} />
+
+# MeetingCard
+
+### Displays a card
```
Could you remove or change it to `# Footer`?

tohuynh replied:
These two lines actually show up in the docs panel, so need to change accordingly.

![image](https://user-images.githubusercontent.com/37560480/112571625-b0c8c500-8da5-11eb-90f2-b609523334fd.png)

andrewlubrino replied:
@tohuynh Just made these changes. My [gh-pages](https://andrewlubrino.github.io/cdp-frontend/?path=/docs/library-layout-footer--footer) is also updated. 

I only added the documentation comments to `footerLinksSections`. I tried adding documentation comments to other objects that are used indirectly in the `Footer` props (so `links`, `footerLinksSectionName`, and then the members that comprise the `Link` interface, `label` and `url`), but it wasn't showing up in the storybook docs, so I took them out for now. 

Let me know if this works.

tohuynh replied:
Looks good. I think the comments only show for component's direct props.

tohuynh commented on src/components/Layout/Footer/Footer.tsx:
```diff
@@ -0,0 +1,83 @@
+}
+
+export interface FooterProps {
+  /*Array of footer columns, each with a name to be displayed at 
+  the top of the column, and an array of links to be displayed 
+  below the name*/
+  footerLinksSections: FooterLinksSection[];
```
Could you change it to /**<comment>*/ so it shows up in the storybook docs panel?

tohuynh reviewed (approved):
I have just two minor requests but looks good.

JacksonMaxfield commented:
Thanks @andrewlubrino! Merged
